File: smali_generator/app/src/main/java/com/smali_generator/patches/AbProps.java
package com.smali_generator.patches;

import android.util.Log;

import com.arthooks.ArtHooks;

import com.smali_generator.Hook;
import com.smali_generator.HookCategory;
import com.smali_generator.abprops.AbPropStore;
import com.smali_generator.ui.AbPropsActivity;

import org.json.JSONObject;

import java.lang.reflect.Executable;
import java.lang.reflect.Method;
import java.lang.reflect.Modifier;
import java.util.Arrays;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

/**
 * Lets the app's A/B properties be read and changed.
 *
 * A property is a number. The app asks "what is 33604?" and gets back a
 * boolean, an int, a float, a string or a JSON object depending on which of
 * five accessors it called, and that answer is what decides whether a feature
 * exists for this account. Meta turns them on per account from the server, so a
 * feature can be in the build for months before it is switched on -- and
 * answering one of those questions differently is how you find out what it
 * does.
 *
 * The hook is not on those accessors but on the one private method all five
 * end in, which takes the id and returns the value boxed. The accessors are 8
 * dex code units each: a warm profile's inline caches let dex2oat copy one into
 * a hot call site behind a class check, and the override then silently stopped
 * applying there -- a read of property 16520 on 2.26.37.74 under the phone's
 * own speed-profile. The funnel is some two hundred instructions; no compiler
 * filter inlines it, so every read reaches it however the app was compiled.
 *
 * What the funnel does not know is which accessor asked. That is decided by the
 * type the app's own answer comes back as: the caller casts the funnel's result
 * to Boolean, Number, String or JSONObject, so an override is only ever handed
 * back in the shape the original answer had. It costs the original call on an
 * overridden read, which the accessor hooks used to skip, and buys the one
 * thing that cannot be got any other way: a string property and a JSON one
 * store their overrides as the same text, and handing a String to the JSON
 * accessor's cast would crash the app inside its own code.
 *
 * Reached from 13,336 sites on 2.26.36.71, so this is as hot a path as the
 * patch touches, and off by default for that reason. It is a tool for poking at
 * the app, not something a normal session should be carrying.
 *
 * One subclass of the properties class overrides all five accessors -- the
 * properties object used during registration and Drive restore, 51 references
 * against 635 for the one the rest of the app holds. Its reads reach the funnel
 * too, on its own receiver, and are passed through untouched: those are not
 * paths with features to find in them, and it must not become the object the
 * default tables are read from.
 */
public class AbProps implements Hook {

    private static final String TAG = "PATCH";

    /** The boolean accessor. A receiver whose class declares it itself is the
     *  registration object above, and is left alone. */
    private static final String BOOL_ACCESSOR = "{{AB_PROPS_BOOL_METHOD_NAME}}";

    private static Class<?> owner;

    /** The receiver class every ordinary read comes from, checked before the map. */
    private static volatile Class<?> lastCovered;

    /** Whether reads on a receiver of this class are this hook's business. A
     *  handful of entries: the properties class has three subclasses. */
    private static final Map<Class<?>, Boolean> covered = new ConcurrentHashMap<>();

    // One replacement and backup per number of reference parameters ahead of
    // the id: four on 2.26.17.72, five since. The layout has to match the
    // target's exactly, and Object stands in for every obfuscated type.

    /** native on purpose: a body here would be inlined into the hook and the backup would
     *  silently answer for the original. ArtHooks rewrites its entry point. */
    static native Object funnel3_backup(Object thiz, Object a, Object b, Object c, int id);

    static Object funnel3_hook(Object thiz, Object a, Object b, Object c, int id) {
        return answer(thiz, id, funnel3_backup(thiz, a, b, c, id));
    }

    /** native on purpose: a body here would be inlined into the hook and the backup would
     *  silently answer for the original. ArtHooks rewrites its entry point. */
    static native Object funnel4_backup(Object thiz, Object a, Object b, Object c, Object d, int id);

    static Object funnel4_hook(Object thiz, Object a, Object b, Object c, Object d, int id) {
        return answer(thiz, id, funnel4_backup(thiz, a, b, c, d, id));
    }

    /** native on purpose: a body here would be inlined into the hook and the backup would
     *  silently answer for the original. ArtHooks rewrites its entry point. */
    static native Object funnel5_backup(Object thiz, Object a, Object b, Object c, Object d, Object e,
                                        int id);

    static Object funnel5_hook(Object thiz, Object a, Object b, Object c, Object d, Object e, int id) {
        return answer(thiz, id, funnel5_backup(thiz, a, b, c, d, e, id));
    }

    /** native on purpose: a body here would be inlined into the hook and the backup would
     *  silently answer for the original. ArtHooks rewrites its entry point. */
    static native Object funnel6_backup(Object thiz, Object a, Object b, Object c, Object d, Object e,
                                        Object f, int id);

    static Object funnel6_hook(Object thiz, Object a, Object b, Object c, Object d, Object e, Object f,
                               int id) {
        return answer(thiz, id, funnel6_backup(thiz, a, b, c, d, e, f, id));
    }

    /**
     * What the app is told property {@code id} is, given what it would have
     * been told.
     *
     * Never throws: this runs inside every property read the app makes, on
     * whatever thread got there first. With nothing overridden it costs a
     * class compare, a volatile read and the observation.
     */
    private static Object answer(Object thiz, int id, Object value) {
        try {
            if (!covers(thiz)) {
                return value;
            }
            // Before the override: what the screen shows as the app's own
            // value has to be the app's own, not this patch's answer read back.
            AbPropStore.observe(thiz, id, value instanceof JSONObject ? value.toString() : value);
            AbPropStore.Override override = AbPropStore.override(id);
            if (override == null) {
                return value;
            }
            Object shaped = shape(id, override.value, value);
            if (shaped == null) {
                return value;
            }
            override.served = true;
            return shaped;
        } catch (Throwable t) {
            Log.e(TAG, "AbProps: property " + id + " answered with the app's own value: " + t);
            return value;
        }
    }

    /**
     * The override in the shape the caller will cast to, or null to leave the
     * app's answer alone.
     *
     * The shape is read off the app's own answer, since that is what the
     * accessor's cast was written for. The int and float accessors cast to
     * Number, so either kind of number fits either. With no answer to go by
     * -- the app's value was null -- only a type that cannot be mistaken
     * would be safe, and text can: it is a String to one accessor and a
     * JSONObject to another, and the wrong one is a ClassCastException in the
     * app's own code. So null is left alone, and the override shows as not
     * served.
     */
    static Object shape(int id, Object override, Object original) {
        if (override == null || original == null) {
            return null;
        }
        if (original instanceof Boolean) {
            return override instanceof Boolean ? override : null;
        }
        if (original instanceof Number) {
            return override instanceof Number ? override : null;
        }
        if (original instanceof String) {
            return override instanceof String ? override : null;
        }
        if (original instanceof JSONObject && override instanceof String) {
            // Parsed on each read rather than kept: a shared JSONObject is one
            // the app could edit under us.
            try {
                return new JSONObject((String) override);
            } catch (Throwable t) {
                Log.e(TAG, "AbProps: the override for " + id + " is not JSON, using the app's own value: " + t);
            }
        }
        return null;
    }

    /**
     * Whether reads on this receiver are overridden and observed.
     *
     * The registration object is told apart by what it is rather than by
     * name: it declares the accessors itself, where the object the rest of
     * the app holds inherits them. Decided once per class.
     */
    private static boolean covers(Object thiz) {
        if (thiz == null) {
            return false;
        }
        Class<?> cls = thiz.getClass();
        if (cls == lastCovered) {
            return true;
        }
        Boolean known = covered.get(cls);
        if (known == null) {
            try {
                known = !declaresAccessor(cls);
                Log.i(TAG, "AbProps: reads on " + cls.getName() + (known ? " are covered"
                        : " are left alone, it declares the accessors itself"));
            } catch (Throwable t) {
                // Remembered like an answer, so a class that cannot be looked
                // at costs one log line rather than one per read.
                known = false;
                Log.e(TAG, "AbProps: reads on " + cls.getName() + " are left alone, it could not be inspected: " + t);
            }
            covered.put(cls, known);
        }
        if (known) {
            lastCovered = cls;
        }
        return known;
    }

    /** By exact name and parameters: getDeclaredMethods() would resolve every
     *  signature in an obfuscated class to answer, and can fail on any one. */
    private static boolean declaresAccessor(Class<?> cls) {
        for (Class<?> c = cls; c != null && c != owner; c = c.getSuperclass()) {
            try {
                c.getDeclaredMethod(BOOL_ACCESSOR, int.class);
                return true;
            } catch (NoSuchMethodException ignored) {
                // Inherited here; keep walking up to the properties class.
            }
        }
        return false;
    }

    public String id() {
        return "ab_props";
    }

    public String title() {
        return "A/B property overrides";
    }

    public String description() {
        return "Read and change the app's own feature flags.";
    }

    public HookCategory category() {
        return HookCategory.DEVELOPER;
    }

    /** Off unless asked for: a hook on the app's hottest read path, in aid of
     *  a screen most sessions have no use for. */
    public boolean defaultEnabled() {
        return false;
    }

    public String configSummary() {
        // Read here as well as at load: this is called to draw the settings
        // screen, which is reachable with the hook switched off.
        AbPropStore.load();
        int count = AbPropStore.count();
        return count == 0 ? "Browse and override properties"
                : count == 1 ? "1 property overridden" : count + " properties overridden";
    }

    public Class<?> configScreen() {
        return AbPropsActivity.class;
    }

    public void load() {
        // Before anything is hooked: a read that arrived between the redirect
        // and the overrides would answer with the app's own value.
        AbPropStore.load();
        try {
            owner = Class.forName("{{AB_PROPS_CLASS_NAME}}");
            Executable funnel = ArtHooks.find_function(owner,
                    "{{AB_PROPS_FUNNEL_METHOD_NAME}}", "{{AB_PROPS_FUNNEL_METHOD_SIG}}");
            if (funnel == null) {
                Log.e(TAG, "AbProps: " + owner.getName() + " has no funnel of the shape the finder saw");
                return;
            }
            // R8 makes an instance method static when its receiver goes unused,
            // without touching the descriptor the finder matched. Hooking a
            // static one installs cleanly and then dies on the first call, when
            // the backup re-enters the replacement -- outside every try/catch,
            // with the app already running.
            if (Modifier.isStatic(funnel.getModifiers())) {
                Log.e(TAG, "AbProps: the funnel is static in this build, nothing is hooked");
                return;
            }
            Class<?>[] params = ((Method) funnel).getParameterTypes();
            int references = params.length - 1;
            if (references < 0 || params[references] != int.class
                    || ((Method) funnel).getReturnType() != Object.class) {
                Log.e(TAG, "AbProps: the funnel does not take the id last and return Object: " + funnel);
                return;
            }
            for (int i = 0; i < references; i++) {
                if (params[i].isPrimitive()) {
                    Log.e(TAG, "AbProps: the funnel takes a primitive before the id: " + funnel);
                    return;
                }
            }
            Class<?>[] shape = new Class<?>[references + 2];
            Arrays.fill(shape, Object.class);
            shape[references + 1] = int.class;
            Method replacement;
            Method original;
            try {
                replacement = AbProps.class.getDeclaredMethod("funnel" + references + "_hook", shape);
                original = AbProps.class.getDeclaredMethod("funnel" + references + "_backup", shape);
            } catch (NoSuchMethodException e) {
                Log.e(TAG, "AbProps: the funnel takes " + references
                        + " references before the id, and no replacement has that shape");
                return;
            }
            boolean hooked = ArtHooks.hook_function(funnel, replacement, original);
            Log.i(TAG, "AbProps: hooked the funnel " + owner.getName() + "."
                    + ((Method) funnel).getName() + "{{AB_PROPS_FUNNEL_METHOD_SIG}}, hooked=" + hooked);
        } catch (Throwable t) {
            Log.e(TAG, "AbProps: the properties class was not reached, nothing is hooked: " + t);
        }
    }

    public void unload() {
        Log.i(TAG, "AbProps: Patch unloaded");
    }
}

File: artifactory_generator/ab_props.py
import re

from stitch.artifactory_generator.SimpleArtifactoryFinder import SimpleArtifactoryFinder

from artifactory_generator.smali import CLASS_RE

# What the props class throws when asked for an id it has no default for. It is
# a diagnostic string, so it is not renamed, and exactly one class in the app
# carries it: the abstract base every A/B property read goes through.
ANCHOR_RE = re.compile(r'const-string(?:/jumbo)? [vp]\d+, "Unknown IntField: "')

METHOD_RE = re.compile(
    r'^\.method (?P<modifiers>[\w ]*?)(?P<name>[\w$<>]+)'
    r'\((?P<params>[^)\n]*)\)(?P<ret>[\w/$;\[]+)\n'
    r'(?P<body>.*?)^\.end method$',
    re.MULTILINE | re.DOTALL)

# One accessor per value type, keyed by the return descriptor. The names are
# outputs -- across 2.26.17.72, 2.26.27.85, 2.26.29.74, 2.26.33.76 and
# 2.26.36.71 the boolean one alone was A0n, A0q, A0s, A0t and A0y -- but the
# shapes did not move, and each is the only public instance method taking an
# int and returning that type.
GETTERS = {
    'Z': 'BOOL',
    'I': 'INT',
    'F': 'FLOAT',
    'Ljava/lang/String;': 'STRING',
    'Lorg/json/JSONObject;': 'JSON',
}

# How many calls deep an accessor may hand off inside the class before it
# reaches the funnel. Two in every release examined -- accessor, typed static
# helper, funnel -- except 2.26.33.76's int path, which had an extra instance
# hop; one more than that is margin, not a guess at what comes next.
MAX_HOPS = 3

# ART inlines nothing over 32 dex code units, and every instruction is at least
# one. A funnel with more instructions than that cannot be copied into a call
# site by any compiler filter, which is the whole reason for hooking it rather
# than the accessors.
INLINE_BUDGET = 32

DIRECTIVE_RE = re.compile(r'^\s*(?:$|[.:#])')


def _instructions(body: str) -> int:
    count, in_annotation = 0, False
    for line in body.splitlines():
        stripped = line.strip()
        if stripped.startswith('.annotation'):
            in_annotation = True
        elif stripped.startswith('.end annotation'):
            in_annotation = False
        elif not in_annotation and not DIRECTIVE_RE.match(line):
            count += 1
    return count


class AbPropsFinder(SimpleArtifactoryFinder):
    """Finds the one method every A/B property read ends in.

    An A/B property is a number -- 33604 -- and the app asks for its value by
    that number alone, through one of five accessors on a single abstract
    class: give it an int, get back a boolean, an int, a float, a string or a
    JSON object. Between them they are called from 13,336 sites on 2.26.36.71.

    Those accessors used to be the hook. They are 8 dex code units each, so a
    warm profile's inline caches let dex2oat copy one into a hot call site
    behind a class check, and an override then silently stops applying there --
    seen at a read of property 16520 on 2.26.37.74 under the phone's own
    ``speed-profile``. Each accessor is only a hand-off, though: through a typed
    static helper into a private instance method that takes the property's id
    last and returns it boxed. That funnel is shared by all five, and is two
    hundred instructions long; nothing can inline it.

    So the accessors are still resolved, each by its shape, but only to be
    followed: every one of them has to arrive at the same funnel, or this does
    not fire. The class is pinned by the message it throws for an unknown id.

    Its concrete subclasses supply the default tables, and one of them
    (``X.0Ck`` on 2.26.37.74) overrides all five accessors: the properties
    object used during registration and Drive restore. Its reads reach the
    funnel too, on its own receiver, and are deliberately left alone -- which is
    what the boolean accessor's name is emitted for: the hook recognises that
    receiver by its declaring the accessor itself.

    Verified to converge on one funnel on 2.26.17.72, 2.26.27.85, 2.26.29.74,
    2.26.33.76, 2.26.36.71, 2.26.36.72, 2.26.36.74 and 2.26.37.74 -- four
    different names, and four reference parameters before the id in the oldest
    against five since.
    """

    def __init__(self, args):
        super().__init__(args)
        self.is_once = True
        self.is_found = False

    def class_filter(self, class_data: str) -> bool:
        return ANCHOR_RE.search(class_data) is not None

    def extract_artifacts(self, artifacts: dict, class_data: str) -> None:
        class_match = CLASS_RE.search(class_data)
        if class_match is None:
            return
        descriptor = 'L%s;' % class_match.group('name')

        methods = {(m.group('name'), m.group('params'), m.group('ret')): m
                   for m in METHOD_RE.finditer(class_data)}

        found = {}
        for method in methods.values():
            modifiers = method.group('modifiers')
            if method.group('params') != 'I' or 'public' not in modifiers:
                continue
            # An abstract declaration has no body to follow, and a static one
            # is not what the app calls: every read site is invoke-virtual.
            if 'abstract' in modifiers or 'static' in modifiers:
                continue
            prefix = GETTERS.get(method.group('ret'))
            if prefix is not None:
                found.setdefault(prefix, []).append(method)

        # All five, each unambiguous. A missing or doubled shape means this is
        # not the class it looks like.
        if sorted(found) != sorted(GETTERS.values()):
            return
        if any(len(candidates) != 1 for candidates in found.values()):
            return

        funnels = {self._funnel(descriptor, methods, found[prefix][0]) for prefix in found}
        if len(funnels) != 1 or None in funnels:
            return
        funnel = methods.get(funnels.pop())
        if funnel is None or 'static' in funnel.group('modifiers'):
            return
        if _instructions(funnel.group('body')) <= INLINE_BUDGET:
            return

        artifacts['AB_PROPS_CLASS_NAME'] = class_match.group('name').replace('/', '.')
        artifacts['AB_PROPS_BOOL_METHOD_NAME'] = found['BOOL'][0].group('name')
        artifacts['AB_PROPS_FUNNEL_METHOD_NAME'] = funnel.group('name')
        artifacts['AB_PROPS_FUNNEL_METHOD_SIG'] = '(%s)%s' % (funnel.group('params'), funnel.group('ret'))
        self.is_found = True

    @staticmethod
    def _funnel(descriptor, methods, accessor):
        """The funnel this accessor hands off to, or None when it does not
        reach exactly one.

        The funnel is recognised by the call rather than by a list of
        parameter types, which moved between releases: an invoke-direct into
        this class, of a method returning Object whose last parameter is the
        int id. Anything else the class calls on the way -- the typed static
        helpers, an instance hop -- is followed, up to MAX_HOPS deep.
        """
        call_re = re.compile(
            r'invoke-(?P<kind>static|direct)(?:/range)? \{[^}]*\}, %s->'
            r'(?P<name>[\w$<>]+)\((?P<params>[^)\n]*)\)(?P<ret>[\w/$;\[]+)' % re.escape(descriptor))
        reached = set()
        frontier = [accessor]
        for _ in range(MAX_HOPS):
            following = []
            for method in frontier:
                for call in call_re.finditer(method.group('body')):
                    key = (call.group('name'), call.group('params'), call.group('ret'))
                    if (call.group('kind') == 'direct' and key[2] == 'Ljava/lang/Object;'
                            and key[1].endswith('I') and not key[1].endswith('[I')):
                        reached.add(key)
                    elif key in methods and key[0] != '<init>':
                        following.append(methods[key])
            if reached:
                break
            frontier = following
        return reached.pop() if len(reached) == 1 else None
